### ISSUE TITLE
Broken javascript: no AMD modules 

### ISSUE THREAD

**Issue by moodleulpgc** [turnitin/moodle-plagiarism_turnitin#416]
Hi:
We have installed mod_turnitintooltwo & plagiarism_turnitin (latest versions) in moodle 3.5  with boost theme.

When the plugin shows in assign of forum submitted files, javascript is broken for all page. No other js element woks (for instance, the main menu). 
I have traced the problem to plagiarism/turnitin/lib.php function load_page_components() 
Loading jquery-3.3.1.min.js and other modules evoke an empty definition error 

In my experience this happens in moodle when mixing other plugins that also use javascript. Moodle now REQUIRES js to be encapsulated in an AMD module, otherwise it may conflict with other plugins. 

Since this blocks standard page navigation and Turnitin must be usable with other assign or forum plugins, this should be considered a bug here and demands correction.

First check if jquery o other js are loaded from other plugins, and if needed lad as an AMD module.

Thanks

**Comment by mkpelletier** [comment]
Please refer to the javascript console for this:

[Error] Error: Mismatched anonymous define() module: function (n){"use strict";var a=function(e){return!1===e||""===e||null==e},i=function(e){return 0<n(e).length},o=function(e){return"string"==typeof e||e instanceof String},r=function(e,t){return-1!==n.inArray(e,t)},f=function(e,t){this._id=f._idCounter++,this.element=n(e).addClass("iconpicker-element"),this._trigger("iconpickerCreate"),this.options=n.extend({},f.defaultOptions,this.element.data(),t),this.options.templates=n.extend({},f.defaultOptions.templates,this.options.templates),this.options.originalPlacement=this.options.placement,this.container=!!i(this.options.container)&&n(this.options.container),!1===this.container&&(this.element.is(".dropdown-toggle")?this.container=n("~ .dropdown-menu:first",this.element):this.container=this.element.is("input,textarea,button,.btn")?this.element.parent():this.element),this.container.addClass("iconpicker-container"),this.isDropdownMenu()&&(this.options.templates.search=!1,this.options.templates.buttons=!1,this.options.placement="inline"),this.input=!!this.element.is("input,textarea")&&this.element.addClass("iconpicker-input"),!1===this.input&&(this.input=this.container.find(this.options.input),this.input.is("input,textarea")||(this.input=!1)),this.component=this.isDropdownMenu()?this.container.parent().find(this.options.component):this.container.find(this.options.component),0===this.component.length?this.component=!1:this.component.find("i").addClass("iconpicker-component"),this._createPopover(),this._createIconpicker(),0===this.getAcceptButton().length&&(this.options.mustAccept=!1),this.isInputGroup()?this.container.parent().append(this.popover):this.container.append(this.popover),this._bindElementEvents(),this._bindWindowEvents(),this.update(this.options.selected),this.isInline()&&this.show(),this._trigger("iconpickerCreated")};f._idCounter=0,f.defaultOptions={title:!1,selected:!1,defaultValue:!1,placement:"bottom",collision:"none",animation:!0,hideOnSelect:!1,showFooter:!1,searchInFooter:!1,mustAccept:!1,selectedCustomClass:"bg-primary",icons:[],fullClassFormatter:function(e){return"fa "+e},input:"input,.iconpicker-input",inputSearch:!1,container:!1,component:".input-group-addon,.iconpicker-component",templates:{popover:'<div class="iconpicker-popover popover"><div class="arrow"></div><div class="popover-title"></div><div class="popover-content"></div></div>',footer:'<div class="popover-footer"></div>',buttons:'<button class="iconpicker-btn iconpicker-btn-cancel btn btn-default btn-sm">Cancel</button> <button class="iconpicker-btn iconpicker-btn-accept btn btn-primary btn-sm">Accept</button>',search:'<input type="search" class="form-control iconpicker-search" placeholder="Type to filter" />',iconpicker:'<div class="iconpicker"><div class="iconpicker-items"></div></div>',iconpickerItem:'<a role="button" href="#" class="iconpicker-item"><i></i></a>'}},f.batch=function(e,t){var a=Array.prototype.slice.call(arguments,2);return n(e).each(function(){var e=n(this).data("iconpicker");e&&e[t].apply(e,a)})},f.prototype={constructor:f,options:{},_id:0,_trigger:function(e,t){t=t||{},this.element.trigger(n.extend({type:e,iconpickerInstance:this},t))},_createPopover:function(){this.popover=n(this.options.templates.popover);var e=this.popover.find(".popover-title");if(this.options.title&&e.append(n('<div class="popover-title-text">'+this.options.title+"</div>")),this.hasSeparatedSearchInput()&&!this.options.searchInFooter?e.append(this.options.templates.search):this.options.title||e.remove(),this.options.showFooter&&!a(this.options.templates.footer)){var t=n(this.options.templates.footer);this.hasSeparatedSearchInput()&&this.options.searchInFooter&&t.append(n(this.options.templates.search)),a(this.options.templates.buttons)||t.append(n(this.options.templates.buttons)),this.popover.append(t)}return!0===this.options.animation&&this.popover.addClass("fade"),this.popover},_createIconpicker:function(){var a=this;this.iconpicker=n(this.options.templates.iconpicker);var e=function(e){var t=n(this);return t.is("i")&&(t=t.parent()),a._trigger("iconpickerSelect",{iconpickerItem:t,iconpickerValue:a.iconpickerValue}),!1===a.options.mustAccept?(a.update(t.data("iconpickerValue")),a._trigger("iconpickerSelected",{iconpickerItem:this,iconpickerValue:a.iconpickerValue})):a.update(t.data("iconpickerValue"),!0),a.options.hideOnSelect&&!1===a.options.mustAccept&&a.hide(),e.preventDefault(),!1};for(var t in this.options.icons)if("string"==typeof this.options.icons[t]){var i=n(this.options.templates.iconpickerItem);i.find("i").addClass(this.options.fullClassFormatter(this.options.icons[t])),i.data("iconpickerValue",this.options.icons[t]).on("click.iconpicker",e),this.iconpicker.find(".iconpicker-items").append(i.attr("title","."+this.options.icons[t]))}return this.popover.find(".popover-content").append(this.iconpicker),this.iconpicker},_isEventInsideIconpicker:function(e){var t=n(e.target);return!((!t.hasClass("iconpicker-element")||t.hasClass("iconpicker-element")&&!t.is(this.element))&&0===t.parents(".iconpicker-popover").length)},_bindElementEvents:function(){var t=this;this.getSearchInput().on("keyup.iconpicker",function(){t.filter(n(this).val().toLowerCase())}),this.getAcceptButton().on("click.iconpicker",function(){var e=t.iconpicker.find(".iconpicker-selected").get(0);t.update(t.iconpickerValue),t._trigger("iconpickerSelected",{iconpickerItem:e,iconpickerValue:t.iconpickerValue}),t.isInline()||t.hide()}),this.getCancelButton().on("click.iconpicker",function(){t.isInline()||t.hide()}),this.element.on("focus.iconpicker",function(e){t.show(),e.stopPropagation()}),this.hasComponent()&&this.component.on("click.iconpicker",function(){t.toggle()}),this.hasInput()&&this.input.on("keyup.iconpicker",function(e){r(e.keyCode,[38,40,37,39,16,17,18,9,8,91,93,20,46,186,190,46,78,188,44,86])?t._updateFormGroupStatus(!1!==t.getValid(this.value)):t.update(),!0===t.options.inputSearch&&t.filter(n(this).val().toLowerCase())})},_bindWindowEvents:function(){var e=n(window.document),t=this,a=".iconpicker.inst"+this._id;return n(window).on("resize.iconpicker"+a+" orientationchange.iconpicker"+a,function(e){t.popover.hasClass("in")&&t.updatePlacement()}),t.isInline()||e.on("mouseup"+a,function(e){return t._isEventInsideIconpicker(e)||t.isInline()||t.hide(),e.stopPropagation(),e.preventDefault(),!1}),!1},_unbindElementEvents:function(){this.popover.off(".iconpicker"),this.element.off(".iconpicker"),this.hasInput()&&this.input.off(".iconpicker"),this.hasComponent()&&this.component.off(".iconpicker"),this.hasContainer()&&this.container.off(".iconpicker")},_unbindWindowEvents:function(){n(window).off(".iconpicker.inst"+this._id),n(window.document).off(".iconpicker.inst"+this._id)},updatePlacement:function(e,t){e=e||this.options.placement,this.options.placement=e,t=!0===(t=t||this.options.collision)?"flip":t;var a={at:"right bottom",my:"right top",of:this.hasInput()&&!this.isInputGroup()?this.input:this.container,collision:!0===t?"flip":t,within:window};if(this.popover.removeClass("inline topLeftCorner topLeft top topRight topRightCorner rightTop right rightBottom bottomRight bottomRightCorner bottom bottomLeft bottomLeftCorner leftBottom left leftTop"),"object"==typeof e)return this.popover.pos(n.extend({},a,e));switch(e){case"inline":a=!1;break;case"topLeftCorner":a.my="right bottom",a.at="left top";break;case"topLeft":a.my="left bottom",a.at="left top";break;case"top":a.my="center bottom",a.at="center top";break;case"topRight":a.my="right bottom",a.at="right top";break;case"topRightCorner":a.my="left bottom",a.at="right top";break;case"rightTop":a.my="left bottom",a.at="right center";break;case"right":a.my="left center",a.at="right center";break;case"rightBottom":a.my="left top",a.at="right center";break;case"bottomRightCorner":a.my="left top",a.at="right bottom";break;case"bottomRight":a.my="right top",a.at="right bottom";break;case"bottom":a.my="center top",a.at="center bottom";break;case"bottomLeft":a.my="left top",a.at="left bottom";break;case"bottomLeftCorner":a.my="right top",a.at="left bottom";break;case"leftBottom":a.my="right top",a.at="left center";break;case"left":a.my="right center",a.at="left center";break;case"leftTop":a.my="right bottom",a.at="left center";break;default:return!1}return this.popover.css({display:"inline"===this.options.placement?"":"block"}),!1!==a?this.popover.pos(a).css("maxWidth",n(window).width()-this.container.offset().left-5):this.popover.css({top:"auto",right:"auto",bottom:"auto",left:"auto",maxWidth:"none"}),this.popover.addClass(this.options.placement),!0},_updateComponents:function(){if(this.iconpicker.find(".iconpicker-item.iconpicker-selected").removeClass("iconpicker-selected "+this.options.selectedCustomClass),this.iconpickerValue&&this.iconpicker.find("."+this.options.fullClassFormatter(this.iconpickerValue).replace(/ /g,".")).parent().addClass("iconpicker-selected "+this.options.selectedCustomClass),this.hasComponent()){var e=this.component.find("i");0<e.length?e.attr("class",this.options.fullClassFormatter(this.iconpickerValue)):this.component.html(this.getHtml())}},_updateFormGroupStatus:function(e){return!!this.hasInput()&&(!1!==e?this.input.parents(".form-group:first").removeClass("has-error"):this.input.parents(".form-group:first").addClass("has-error"),!0)},getValid:function(e){o(e)||(e="");var t=""===e;return e=n.trim(e),!(!r(e,this.options.icons)&&!t)&&e},setValue:function(e){var t=this.getValid(e);return!1!==t?(this.iconpickerValue=t,this._trigger("iconpickerSetValue",{iconpickerValue:t}),this.iconpickerValue):(this._trigger("iconpickerInvalid",{iconpickerValue:e}),!1)},getHtml:function(){return'<i class="'+this.options.fullClassFormatter(this.iconpickerValue)+'"></i>'},setSourceValue:function(e){return!1!==(e=this.setValue(e))&&""!==e&&(this.hasInput()?this.input.val(this.iconpickerValue):this.element.data("iconpickerValue",this.iconpickerValue),this._trigger("iconpickerSetSourceValue",{iconpickerValue:e})),e},getSourceValue:function(e){var t=e=e||this.options.defaultValue;return void 0!==(t=this.hasInput()?this.input.val():this.element.data("iconpickerValue"))&&""!==t&&null!==t&&!1!==t||(t=e),t},hasInput:function(){return!1!==this.input},isInputSearch:function(){return this.hasInput()&&!0===this.options.inputSearch},isInputGroup:function(){return this.container.is(".input-group")},isDropdownMenu:function(){return this.container.is(".dropdown-menu")},hasSeparatedSearchInput:function(){return!1!==this.options.templates.search&&!this.isInputSearch()},hasComponent:function(){return!1!==this.component},hasContainer:function(){return!1!==this.container},getAcceptButton:function(){return this.popover.find(".iconpicker-btn-accept")},getCancelButton:function(){return this.popover.find(".iconpicker-btn-cancel")},getSearchInput:function(){return this.popover.find(".iconpicker-search")},filter:function(i){if(a(i))return this.iconpicker.find(".iconpicker-item").show(),n(!1);var o=[];return this.iconpicker.find(".iconpicker-item").each(function(){var e=n(this),t=e.attr("title").toLowerCase(),a=!1;try{a=new RegExp(i,"g")}catch(e){a=!1}!1!==a&&t.match(a)?(o.push(e),e.show()):e.hide()}),o},show:function(){if(this.popover.hasClass("in"))return!1;n.iconpicker.batch(n(".iconpicker-popover.in:not(.inline)").not(this.popover),"hide"),this._trigger("iconpickerShow"),this.updatePlacement(),this.popover.addClass("in"),setTimeout(n.proxy(function(){this.popover.css("display",this.isInline()?"":"block"),this._trigger("iconpickerShown")},this),this.options.animation?300:1)},hide:function(){if(!this.popover.hasClass("in"))return!1;this._trigger("iconpickerHide"),this.popover.removeClass("in"),setTimeout(n.proxy(function(){this.popover.css("display","none"),this.getSearchInput().val(""),this.filter(""),this._trigger("iconpickerHidden")},this),this.options.animation?300:1)},toggle:function(){this.popover.is(":visible")?this.hide():this.show(!0)},update:function(e,t){return e=e||this.getSourceValue(this.iconpickerValue),this._trigger("iconpickerUpdate"),!0===t?e=this.setValue(e):(e=this.setSourceValue(e),this._updateFormGroupStatus(!1!==e)),!1!==e&&this._updateComponents(),this._trigger("iconpickerUpdated"),e},destroy:function(){this._trigger("iconpickerDestroy"),this.element.removeData("iconpicker").removeData("iconpickerValue").removeClass("iconpicker-element"),this._unbindElementEvents(),this._unbindWindowEvents(),n(this.popover).remove(),this._trigger("iconpickerDestroyed")},disable:function(){return!!this.hasInput()&&(this.input.prop("disabled",!0),!0)},enable:function(){return!!this.hasInput()&&(this.input.prop("disabled",!1),!0)},isDisabled:function(){return!!this.hasInput()&&!0===this.input.prop("disabled")},isInline:function(){return"inline"===this.options.placement||this.popover.hasClass("inline")}},n.iconpicker=f,n.fn.iconpicker=function(t){return this.each(function(){var e=n(this);e.data("iconpicker")||e.data("iconpicker",new f(this,"object"==typeof t?t:{}))})},f.defaultOptions.icons=["fa-500px","fa-address-book","fa-address-book-o","fa-address-card","fa-address-card-o","fa-adjust","fa-adn","fa-align-center","fa-align-justify","fa-align-left","fa-align-right","fa-amazon","fa-ambulance","fa-american-sign-language-interpreting","fa-anchor","fa-android","fa-angellist","fa-angle-double-down","fa-angle-double-left","fa-angle-double-right","fa-angle-double-up","fa-angle-down","fa-angle-left","fa-angle-right","fa-angle-up","fa-apple","fa-archive","fa-area-chart","fa-arrow-circle-down","fa-arrow-circle-left","fa-arrow-circle-o-down","fa-arrow-circle-o-left","fa-arrow-circle-o-right","fa-arrow-circle-o-up","fa-arrow-circle-right","fa-arrow-circle-up","fa-arrow-down","fa-arrow-left","fa-arrow-right","fa-arrow-up","fa-arrows","fa-arrows-alt","fa-arrows-h","fa-arrows-v","fa-asl-interpreting","fa-assistive-listening-systems","fa-asterisk","fa-at","fa-audio-description","fa-automobile","fa-backward","fa-balance-scale","fa-ban","fa-bandcamp","fa-bank","fa-bar-chart","fa-bar-chart-o","fa-barcode","fa-bars","fa-bath","fa-bathtub","fa-battery","fa-battery-0","fa-battery-1","fa-battery-2","fa-battery-3","fa-battery-4","fa-battery-empty","fa-battery-full","fa-battery-half","fa-battery-quarter","fa-battery-three-quarters","fa-bed","fa-beer","fa-behance","fa-behance-square","fa-bell","fa-bell-o","fa-bell-slash","fa-bell-slash-o","fa-bicycle","fa-binoculars","fa-birthday-cake","fa-bitbucket","fa-bitbucket-square","fa-bitcoin","fa-black-tie","fa-blind","fa-bluetooth","fa-bluetooth-b","fa-bold","fa-bolt","fa-bomb","fa-book","fa-bookmark","fa-bookmark-o","fa-braille","fa-briefcase","fa-btc","fa-bug","fa-building","fa-building-o","fa-bullhorn","fa-bullseye","fa-bus","fa-buysellads","fa-cab","fa-calculator","fa-calendar","fa-calendar-check-o","fa-calendar-minus-o","fa-calendar-o","fa-calendar-plus-o","fa-calendar-times-o","fa-camera","fa-camera-retro","fa-car","fa-caret-down","fa-caret-left","fa-caret-right","fa-caret-square-o-down","fa-caret-square-o-left","fa-caret-square-o-right","fa-caret-square-o-up","fa-caret-up","fa-cart-arrow-down","fa-cart-plus","fa-cc","fa-cc-amex","fa-cc-diners-club","fa-cc-discover","fa-cc-jcb","fa-cc-mastercard","fa-cc-paypal","fa-cc-stripe","fa-cc-visa","fa-certificate","fa-chain","fa-chain-broken","fa-check","fa-check-circle","fa-check-circle-o","fa-check-square","fa-check-square-o","fa-chevron-circle-down","fa-chevron-circle-left","fa-chevron-circle-right","fa-chevron-circle-up","fa-chevron-down","fa-chevron-left","fa-chevron-right","fa-chevron-up","fa-child","fa-chrome","fa-circle","fa-circle-o","fa-circle-o-notch","fa-circle-thin","fa-clipboard","fa-clock-o","fa-clone","fa-close","fa-cloud","fa-cloud-download","fa-cloud-upload","fa-cny","fa-code","fa-code-fork","fa-codepen","fa-codiepie","fa-coffee","fa-cog","fa-cogs","fa-columns","fa-comment","fa-comment-o","fa-commenting","fa-commenting-o","fa-comments","fa-comments-o","fa-compass","fa-compress","fa-connectdevelop","fa-contao","fa-copy","fa-copyright","fa-creative-commons","fa-credit-card","fa-credit-card-alt","fa-crop","fa-crosshairs","fa-css3","fa-cube","fa-cubes","fa-cut","fa-cutlery","fa-dashboard","fa-dashcube","fa-database","fa-deaf","fa-deafness","fa-dedent","fa-delicious","fa-desktop","fa-deviantart","fa-diamond","fa-digg","fa-dollar","fa-dot-circle-o","fa-download","fa-dribbble","fa-drivers-license","fa-drivers-license-o","fa-dropbox","fa-drupal","fa-edge","fa-edit","fa-eercast","fa-eject","fa-ellipsis-h","fa-ellipsis-v","fa-empire","fa-envelope","fa-envelope-o","fa-envelope-open","fa-envelope-open-o","fa-envelope-square","fa-envira","fa-eraser","fa-etsy","fa-eur","fa-euro","fa-exchange","fa-exclamation","fa-exclamation-circle","fa-exclamation-triangle","fa-expand","fa-expeditedssl","fa-external-link","fa-external-link-square","fa-eye","fa-eye-slash","fa-eyedropper","fa-fa","fa-facebook","fa-facebook-f","fa-facebook-official","fa-facebook-square","fa-fast-backward","fa-fast-forward","fa-fax","fa-feed","fa-female","fa-fighter-jet","fa-file","fa-file-archive-o","fa-file-audio-o","fa-file-code-o","fa-file-excel-o","fa-file-image-o","fa-file-movie-o","fa-file-o","fa-file-pdf-o","fa-file-photo-o","fa-file-picture-o","fa-file-powerpoint-o","fa-file-sound-o","fa-file-text","fa-file-text-o","fa-file-video-o","fa-file-word-o","fa-file-zip-o","fa-files-o","fa-film","fa-filter","fa-fire","fa-fire-extinguisher","fa-firefox","fa-first-order","fa-flag","fa-flag-checkered","fa-flag-o","fa-flash","fa-flask","fa-flickr","fa-floppy-o","fa-folder","fa-folder-o","fa-folder-open","fa-folder-open-o","fa-font","fa-font-awesome","fa-fonticons","fa-fort-awesome","fa-forumbee","fa-forward","fa-foursquare","fa-free-code-camp","fa-frown-o","fa-futbol-o","fa-gamepad","fa-gavel","fa-gbp","fa-ge","fa-gear","fa-gears","fa-genderless","fa-get-pocket","fa-gg","fa-gg-circle","fa-gift","fa-git","fa-git-square","fa-github","fa-github-alt","fa-github-square","fa-gitlab","fa-gittip","fa-glass","fa-glide","fa-glide-g","fa-globe","fa-google","fa-google-plus","fa-google-plus-circle","fa-google-plus-official","fa-google-plus-square","fa-google-wallet","fa-graduation-cap","fa-gratipay","fa-grav","fa-group","fa-h-square","fa-hacker-news","fa-hand-grab-o","fa-hand-lizard-o","fa-hand-o-down","fa-hand-o-left","fa-hand-o-right","fa-hand-o-up","fa-hand-paper-o","fa-hand-peace-o","fa-hand-pointer-o","fa-hand-rock-o","fa-hand-scissors-o","fa-hand-spock-o","fa-hand-stop-o","fa-handshake-o","fa-hard-of-hearing","fa-hashtag","fa-hdd-o","fa-header","fa-headphones","fa-heart","fa-heart-o","fa-heartbeat","fa-history","fa-home","fa-hospital-o","fa-hotel","fa-hourglass","fa-hourglass-1","fa-hourglass-2","fa-hourglass-3","fa-hourglass-end","fa-hourglass-half","fa-hourglass-o","fa-hourglass-start","fa-houzz","fa-html5","fa-i-cursor","fa-id-badge","fa-id-card","fa-id-card-o","fa-ils","fa-image","fa-imdb","fa-inbox","fa-indent","fa-industry","fa-info","fa-info-circle","fa-inr","fa-instagram","fa-institution","fa-internet-explorer","fa-intersex","fa-ioxhost","fa-italic","fa-joomla","fa-jpy","fa-jsfiddle","fa-key","fa-keyboard-o","fa-krw","fa-language","fa-laptop","fa-lastfm","fa-lastfm-square","fa-leaf","fa-leanpub","fa-legal","fa-lemon-o","fa-level-down","fa-level-up","fa-life-bouy","fa-life-buoy","fa-life-ring","fa-life-saver","fa-lightbulb-o","fa-line-chart","fa-link","fa-linkedin","fa-linkedin-square","fa-linode","fa-linux","fa-list","fa-list-alt","fa-list-ol","fa-list-ul","fa-location-arrow","fa-lock","fa-long-arrow-down","fa-long-arrow-left","fa-long-arrow-right","fa-long-arrow-up","fa-low-vision","fa-magic","fa-magnet","fa-mail-forward","fa-mail-reply","fa-mail-reply-all","fa-male","fa-map","fa-map-marker","fa-map-o","fa-map-pin","fa-map-signs","fa-mars","fa-mars-double","fa-mars-stroke","fa-mars-stroke-h","fa-mars-stroke-v","fa-maxcdn","fa-meanpath","fa-medium","fa-medkit","fa-meetup","fa-meh-o","fa-mercury","fa-microchip","fa-microphone","fa-microphone-slash","fa-minus","fa-minus-circle","fa-minus-square","fa-minus-square-o","fa-mixcloud","fa-mobile","fa-mobile-phone","fa-modx","fa-money","fa-moon-o","fa-mortar-board","fa-motorcycle","fa-mouse-pointer","fa-music","fa-navicon","fa-neuter","fa-newspaper-o","fa-object-group","fa-object-ungroup","fa-odnoklassniki","fa-odnoklassniki-square","fa-opencart","fa-openid","fa-opera","fa-optin-monster","fa-outdent","fa-pagelines","fa-paint-brush","fa-paper-plane","fa-paper-plane-o","fa-paperclip","fa-paragraph","fa-paste","fa-pause","fa-pause-circle","fa-pause-circle-o","fa-paw","fa-paypal","fa-pencil","fa-pencil-square","fa-pencil-square-o","fa-percent","fa-phone","fa-phone-square","fa-photo","fa-picture-o","fa-pie-chart","fa-pied-piper","fa-pied-piper-alt","fa-pied-piper-pp","fa-pinterest","fa-pinterest-p","fa-pinterest-square","fa-plane","fa-play","fa-play-circle","fa-play-circle-o","fa-plug","fa-plus","fa-plus-circle","fa-plus-square","fa-plus-square-o","fa-podcast","fa-power-off","fa-print","fa-product-hunt","fa-puzzle-piece","fa-qq","fa-qrcode","fa-question","fa-question-circle","fa-question-circle-o","fa-quora","fa-quote-left","fa-quote-right","fa-ra","fa-random","fa-ravelry","fa-rebel","fa-recycle","fa-reddit","fa-reddit-alien","fa-reddit-square","fa-refresh","fa-registered","fa-remove","fa-renren","fa-reorder","fa-repeat","fa-reply","fa-reply-all","fa-resistance","fa-retweet","fa-rmb","fa-road","fa-rocket","fa-rotate-left","fa-rotate-right","fa-rouble","fa-rss","fa-rss-square","fa-rub","fa-ruble","fa-rupee","fa-s15","fa-safari","fa-save","fa-scissors","fa-scribd","fa-search","fa-search-minus","fa-search-plus","fa-sellsy","fa-send","fa-send-o","fa-server","fa-share","fa-share-alt","fa-share-alt-square","fa-share-square","fa-share-square-o","fa-shekel","fa-sheqel","fa-shield","fa-ship","fa-shirtsinbulk","fa-shopping-bag","fa-shopping-basket","fa-shopping-cart","fa-shower","fa-sign-in","fa-sign-language","fa-sign-out","fa-signal","fa-signing","fa-simplybuilt","fa-sitemap","fa-skyatlas","fa-skype","fa-slack","fa-sliders","fa-slideshare","fa-smile-o","fa-snapchat","fa-snapchat-ghost","fa-snapchat-square","fa-snowflake-o","fa-soccer-ball-o","fa-sort","fa-sort-alpha-asc","fa-sort-alpha-desc","fa-sort-amount-asc","fa-sort-amount-desc","fa-sort-asc","fa-sort-desc","fa-sort-down","fa-sort-numeric-asc","fa-sort-numeric-desc","fa-sort-up","fa-soundcloud","fa-space-shuttle","fa-spinner","fa-spoon","fa-spotify","fa-square","fa-square-o","fa-stack-exchange","fa-stack-overflow","fa-star","fa-star-half","fa-star-half-empty","fa-star-half-full","fa-star-half-o","fa-star-o","fa-steam","fa-steam-square","fa-step-backward","fa-step-forward","fa-stethoscope","fa-sticky-note","fa-sticky-note-o","fa-stop","fa-stop-circle","fa-stop-circle-o","fa-street-view","fa-strikethrough","fa-stumbleupon","fa-stumbleupon-circle","fa-subscript","fa-subway","fa-suitcase","fa-sun-o","fa-superpowers","fa-superscript","fa-support","fa-table","fa-tablet","fa-tachometer","fa-tag","fa-tags","fa-tasks","fa-taxi","fa-telegram","fa-television","fa-tencent-weibo","fa-terminal","fa-text-height","fa-text-width","fa-th","fa-th-large","fa-th-list","fa-themeisle","fa-thermometer","fa-thermometer-0","fa-thermometer-1","fa-thermometer-2","fa-thermometer-3","fa-thermometer-4","fa-thermometer-empty","fa-thermometer-full","fa-thermometer-half","fa-thermometer-quarter","fa-thermometer-three-quarters","fa-thumb-tack","fa-thumbs-down","fa-thumbs-o-down","fa-thumbs-o-up","fa-thumbs-up","fa-ticket","fa-times","fa-times-circle","fa-times-circle-o","fa-times-rectangle","fa-times-rectangle-o","fa-tint","fa-toggle-down","fa-toggle-left","fa-toggle-off","fa-toggle-on","fa-toggle-right","fa-toggle-up","fa-trademark","fa-train","fa-transgender","fa-transgender-alt","fa-trash","fa-trash-o","fa-tree","fa-trello","fa-tripadvisor","fa-trophy","fa-truck","fa-try","fa-tty","fa-tumblr","fa-tumblr-square","fa-turkish-lira","fa-tv","fa-twitch","fa-twitter","fa-twitter-square","fa-umbrella","fa-underline","fa-undo","fa-universal-access","fa-university","fa-unlink","fa-unlock","fa-unlock-alt","fa-unsorted","fa-upload","fa-usb","fa-usd","fa-user","fa-user-circle","fa-user-circle-o","fa-user-md","fa-user-o","fa-user-plus","fa-user-secret","fa-user-times","fa-users","fa-vcard","fa-vcard-o","fa-venus","fa-venus-double","fa-venus-mars","fa-viacoin","fa-viadeo","fa-viadeo-square","fa-video-camera","fa-vimeo","fa-vimeo-square","fa-vine","fa-vk","fa-volume-control-phone","fa-volume-down","fa-volume-off","fa-volume-up","fa-warning","fa-wechat","fa-weibo","fa-weixin","fa-whatsapp","fa-wheelchair","fa-wheelchair-alt","fa-wifi","fa-wikipedia-w","fa-window-close","fa-window-close-o","fa-window-maximize","fa-window-minimize","fa-window-restore","fa-windows","fa-won","fa-wordpress","fa-wpbeginner","fa-wpexplorer","fa-wpforms","fa-wrench","fa-xing","fa-xing-square","fa-y-combinator","fa-y-combinator-square","fa-yahoo","fa-yc","fa-yc-square","fa-yelp","fa-yen","fa-yoast","fa-youtube","fa-youtube-play","fa-youtube-square"]}
http://requirejs.org/docs/errors.html#mismatch
	defaultOnError (require.min.js:1:935)
	c (require.min.js:1:3260)
	v (require.min.js:1:4659)
	o (require.min.js:1:11170)
	requirejs (require.min.js:1:14745)
	(anonymous function) (view.php:3490)
	execCb (require.min.js:1:12867)
	check (require.min.js:1:6622)
	(anonymous function) (require.min.js:1:9042)
	(anonymous function) (require.min.js:1:820)
	(anonymous function) (require.min.js:1:9499)
	each (require.min.js:1:291)
	emit (require.min.js:1:9470)
	check (require.min.js:1:7179)
	enable (require.min.js:1:9369)
	init (require.min.js:1:5728)
	h (require.min.js:1:4292)
	completeLoad (require.min.js:1:12092)
	onScriptLoad (require.min.js:1:13029)

This is a real problem as none of my admin can get their work done.

**Comment by dwinn** [comment]
Hi @moodleulpgc,

I am unable to reproduce this on the latest versions of our plugins. Moving the JS to AMD modules is something that is within our plans, but these errors shouldn't be occurring with the JS as it is.

Our Moodle environments only contain the Turnitin plugins, so I wonder if there are other plugins you have which are causing Turnitin plugins to conflict with, and that is why we are not seeing it.

**Comment by mkpelletier** [comment]
> Our Moodle environments only contain the Turnitin plugins, so I wonder if there are other plugins you have which are causing Turnitin plugins to conflict with, and that is why we are not seeing it.

@dwinn, I think this is the point. A modular system is not very modular if we can only install one plugin (e.g. Turnitin).  Most Moodle installations I have ever seen or managed install third-party plugins, all of which play nicely when they follow the prescribed plugin convention.  This being the prevailing use case, what sense does it make to have a test environment with only one plugin? Surely, it would make sense for your developers to use a test environment that is representative of the predominant use case?  As it is, the Turnitin plagiarism plugin breaks Moodle or at least limits its modular extensibility because it does not follow the specified conventions.

https://docs.moodle.org/dev/Guide_to_adding_third_party_jQuery_for_AMD
https://docs.moodle.org/dev/Javascript_Modules

**Comment by dwinn** [comment]
Hi @mkpelletier,

The reason we aren't using AMD modules already is because they didn't exist in Moodle until version 2.9. I understand that was released a few years ago but it was only recently (September) that we dropped support for Moodle 2.8. That's why we've kept with raw JS for so long but we have plans to change to AMD modules.

My suspicion is that there is more to it than the plugin simply not using AMD modules because with Moodle having them since 2.9 and this (plus #419) being the first time we've heard of this problem it seems a bit odd that this would suddenly start happening. It could be something else that changed recently.

Do you know of a publicly available plugin I can install in which this problem will occur so I can test this issue with?

**Comment by mkpelletier** [comment]
Hi @dwinn,

Thanks for the response.  There are a couple of very popular plugins that should probably be included in any test environment.  [Essential theme](https://github.com/gjb2048/moodle-theme_essential) and Adaptable are the two most popular themes in the Moodle repositories, I believe.  Essential is particularly representative of Moodle coding standards, so this should probably be included in your test environment.

There are other popular plugins, and two in particular appear to conflict with the plagiarism plugin: Annotate PDF and Mahara portfolio submissions.  The first of these provides similar function to Turnitin, but within Moodle's own marking interface, so I am not sure of the use case that would require both to be enabled. However, there seem to be rational use cases where Mahara and Turnitin would both be used on a single assignment.

I hope this helps.

**Comment by tristdac** [comment]
Same exact problem here since update this week to most recent versions of Turnitin and Turnitin plagiarism. Currently on Moodle v3.4.8 and Adaptable theme.

**Comment by mkpelletier** [comment]
This problem needs some urgent attention.  I have become despondent over the quality control issues with Turnitin plugins.  Are there any updates on this yet?

**Comment by dwinn** [comment]
Hi @mkpelletier,

We see this issue as a priority and we are planning to release a new version of PP in the near future which uses AMD modules for the JS. There is a lot of JS to switch over and be tested so we can't give a date for this release. This version also won't require Moodle Direct V2 (turnitintooltwo) to be installed so it is quite a big update.

**Comment by dwinn** [comment]
Hi @mkpelletier, @tristdac, @moodleulpgc,

We have just released a big update to PP (v2019050201) and in it we have migrated the majority of the JS to AMD modules. We've migrated everything that is accessed via the assignment, forum and workshop modules. The only parts of the plugin which are not using AMD modules now are the plugin settings (at admin level) and we plan to migrate this in our next release. I believe the work we've done for this release will resolve this problem so I'll close for now.

**Comment by mkpelletier** [comment]
Hi @dwinn,

Thanks for tagging me for this notification.  I wanted to check #436 prior to beginning with this upgrade as I was worried something else might break. I have tested the install on our development server and it seems to work okay, resolving the javascript errors we were getting.

I can confirm that the plagiarism plugin works for both assignments and workshops.  I did get a php notice in my log related to read receipts, but this may be due to the fact that I am testing on a different url than our production system, but I can't say for sure.  Here it is for your interest:
PHP Notice:  Trying to get property 'id' of non-object in /var/www/moodle/plagiarism/turnitin/lib.php on line 1244

Thanks again for the hard work!

Best Regards,
Mathieu
